lib/Service/ReminderService.php

```diff
@@ -0,0 +1,132 @@
+		protected ChatManager $chatManager,
+	) {
+	}
+
+	public function setReminder(string $userId, string $token, int $messageId, int $timestamp): Reminder {
+		try {
+			$reminder = $this->reminderMapper->findForUserAndMessage($userId, $messageId);
+			$reminder->setDateTime(new \DateTime('@' . $timestamp));
+			$this->reminderMapper->update($reminder);
+
+		} catch (DoesNotExistException) {
+			$reminder = new Reminder();
+			$reminder->setUserId($userId);
+			$reminder->setToken($token);
+			$reminder->setMessageId($messageId);
+			$reminder->setDateTime(new \DateTime('@' . $timestamp));
+			$this->reminderMapper->insert($reminder);
+		}
+
+		return $reminder;
+	}
+
+	/**
+	 * @throws DoesNotExistException
+	 */
+	public function getReminder(string $userId, int $messageId): Reminder {
+		return $this->reminderMapper->findForUserAndMessage($userId, $messageId);
+	}
+
+	public function deleteReminder(string $userId, string $token, int $messageId): void {
+		try {
+			$reminder = $this->reminderMapper->findForUserAndMessage($userId, $messageId);
+			$this->reminderMapper->delete($reminder);
+		} catch (DoesNotExistException) {
+			// When the reminder does not exist anymore, the notification could be there
+			$notification = $this->notificationManager->createNotification();
+			$notification->setApp(Application::APP_ID)
+				->setUser($userId)
+				->setObject('reminder', $token)
+				->setMessage('reminder', [
+					'commentId' => $messageId,
+				]);
+			$this->notificationManager->markProcessed($notification);
+		}
+	}
+
+	public function executeReminders(\DateTime $executeBefore): void {
+		$reminders = $this->reminderMapper->findRemindersToExecute($executeBefore);
+
+		if (empty($reminders)) {
+			return;
+		}
+
+		$shouldFlush = $this->notificationManager->defer();
+
+		$messageIds = [];
+		foreach ($reminders as $reminder) {
+			$messageIds[] = $reminder->getMessageId();
+		}
+		$messageIds = array_unique($messageIds);
+
+		$messages = $this->chatManager->getMessagesById($messageIds);
+
+		foreach ($reminders as $reminder) {
+			if (!isset($messages[$reminder->getMessageId()])) {
+				continue;
+			}
+			$message = $messages[$reminder->getMessageId()];
+
+			$notification = $this->notificationManager->createNotification();
+			$notification->setApp(Application::APP_ID)
+				->setUser($reminder->getUserId())
+				->setObject('reminder', $reminder->getToken())
+				->setDateTime($reminder->getDateTime())
+				->setSubject('reminder', [
+					'token' => $reminder->getToken(),
+					'message' => $reminder->getMessageId(),
+					'userType' => $message->getActorType(),
+					'userId' => $message->getActorId(),
+				])
+				->setMessage('reminder', [
+					'commentId' => $reminder->getMessageId(),
+				]);
+			$this->notificationManager->notify($notification);
+		}
+
+		if ($shouldFlush) {
+			$this->notificationManager->flush();
+		}
+
+		$this->reminderMapper->deleteExecutedReminders($executeBefore);
+	}
```

Comment:
I  think this function should be in a database transaction

Reply:
Why is that?

Reply:
Would still happen for the reminders that were sent before the crash, and I would like to avoid a delete per reminder.
So I guess it's a price we have to pay (and since it's only in programmatic cases I'm fine with it).